### PR TITLE
Add gpu benchmarks for the PBS & multi-bit PBS

### PR DESCRIPTION
**Pull request by agnesLeroy**
closes: _please link all relevant issues_

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by agnesLeroy** [comment]
@slab-ci integer_gpu_bench

**Comment by soonum** [comment]
@slab-ci integer_gpu_bench

**Comment by soonum** [comment]
@slab-ci integer_gpu_bench

**Comment by soonum** [comment]
@slab-ci integer_gpu_bench

File: tfhe/benches/core_crypto/pbs_bench.rs [hunk 1]
```diff
@@ -35,6 +37,29 @@ const SHORTINT_BENCH_PARAMS: [ClassicPBSParameters; 19] = [
...
+const SHORTINT_BENCH_PARAMS: [ClassicPBSParameters; 2] = [
+    // PARAM_MESSAGE_1_CARRY_0_KS_PBS,
+    PARAM_MESSAGE_1_CARRY_1_KS_PBS,
+    // PARAM_MESSAGE_2_CARRY_0_KS_PBS,
+    // PARAM_MESSAGE_2_CARRY_1_KS_PBS,
+    PARAM_MESSAGE_2_CARRY_2_KS_PBS,
+    // PARAM_MESSAGE_3_CARRY_0_KS_PBS,
+    // PARAM_MESSAGE_3_CARRY_2_KS_PBS,
+    // PARAM_MESSAGE_3_CARRY_3_KS_PBS,
+    // PARAM_MESSAGE_4_CARRY_0_KS_PBS,
```
Comment (soonum) [inline tfhe/benches/core_crypto/pbs_bench.rs:50]: Should we make a more thorough selection by either selecting all of the parameters (thus removing the need to gate the constant), or by selecting only some of them and removing all the commented ones ?

**Review by soonum — COMMENTED** [review]
I'm gonna trust you on the PBS benchmark implementation code.
Just a minor comment on parameters constant array.

File: tfhe/benches/core_crypto/pbs_bench.rs (hunk 1 above)
Comment (agnesLeroy) [reply tfhe/benches/core_crypto/pbs_bench.rs:50]: I'll remove the gate, so we bench the same parameter sets on GPU & CPU.

File: tfhe/benches/core_crypto/pbs_bench.rs [hunk 2]
```diff
@@ -547,3 +536,290 @@ fn pbs_throughput<Scalar: UnsignedTorus + CastInto<usize> + Sync + Send + Serial
...
+            let mut out_pbs_ct_gpu =
+                CudaLweCiphertextList::from_lwe_ciphertext(&out_pbs_ct, &stream);
+            let h_indexes = &[Scalar::ZERO];
+            stream.synchronize();
+            let mut d_input_indexes = stream.malloc_async::<Scalar>(1u32);
+            let mut d_output_indexes = stream.malloc_async::<Scalar>(1u32);
+            let mut d_lut_indexes = stream.malloc_async::<Scalar>(1u32);
+            stream.copy_to_gpu_async(&mut d_input_indexes, h_indexes.as_ref());
+            stream.copy_to_gpu_async(&mut d_output_indexes, h_indexes.as_ref());
+            stream.copy_to_gpu_async(&mut d_input_indexes, h_indexes.as_ref());
```
Comment (tmontaigu) [inline tfhe/benches/core_crypto/pbs_bench.rs:632]: I think you'll have to rebase and put that in a unsafe block due to https://github.com/zama-ai/tfhe-rs/commit/11db96d3941328cd18f0d04e9dd68a512d4bef07 being merged

File: tfhe/benches/core_crypto/pbs_bench.rs [hunk 3]
```diff
@@ -547,3 +536,290 @@ fn pbs_throughput<Scalar: UnsignedTorus + CastInto<usize> + Sync + Send + Serial
...
```
Comment (tmontaigu) [inline tfhe/benches/core_crypto/pbs_bench.rs:752]: I think you'll have to rebase and put that in a unsafe block due to https://github.com/zama-ai/tfhe-rs/commit/11db96d3941328cd18f0d04e9dd68a512d4bef07 being merged

File: tfhe/benches/core_crypto/pbs_bench.rs [hunk 4]
```diff
@@ -121,25 +99,36 @@ fn throughput_benchmark_parameters<Scalar: UnsignedInteger>(
...
-                params.grouping_factor,
-            )
-        })
-        .collect()
+        let parameters = if cfg!(feature = "gpu") {
+            vec![
+                PARAM_MULTI_BIT_MESSAGE_1_CARRY_1_GROUP_3_KS_PBS,
+                PARAM_MULTI_BIT_MESSAGE_2_CARRY_2_GROUP_3_KS_PBS,
+                PARAM_MULTI_BIT_MESSAGE_1_CARRY_1_GROUP_4_KS_PBS,
+                // PARAM_MULTI_BIT_MESSAGE_3_CARRY_3_GROUP_3_KS_PBS,
```
Comment (soonum) [inline tfhe/benches/core_crypto/pbs_bench.rs:107]: Do we want to benchmark this parameter set at some point ?

**Review by soonum — COMMENTED** [review]
You've added the new group_4 to keycache, great :+1:

File: tfhe/benches/core_crypto/pbs_bench.rs (hunk 4 above)
Comment (agnesLeroy) [reply tfhe/benches/core_crypto/pbs_bench.rs]: Actually yes. Btw @IceTDrinker this PR introduces the GROUP_4 parameter set for MESSAGE_1 CARRY_1, should I keep it in the PR or remove it?

Comment (IceTDrinker) [reply tfhe/benches/core_crypto/pbs_bench.rs]: no no, no group 4 please

Comment (agnesLeroy) [reply tfhe/benches/core_crypto/pbs_bench.rs]: sure I'll remove it sorry

Comment (IceTDrinker) [reply tfhe/benches/core_crypto/pbs_bench.rs]: no worries ! 🙂

**Review by IceTDrinker — COMMENTED** [review]
Benchmarking code looks ok, was it already checked by David ?

**Comment by agnesLeroy** [comment]
Sir yes sir

**Review by IceTDrinker — APPROVED** [review]
All right let's gooo !

**Pull request merged**